base-compat-batteries/src/Data/Tuple/Compat.hs

```diff
@@ -1,6 +1,23 @@
 {-# LANGUAGE CPP, NoImplicitPrelude, PackageImports #-}
+#if __GLASGOW_HASKELL__ >= 708
+{-# LANGUAGE PatternSynonyms #-}
+#endif
 module Data.Tuple.Compat
-  ( Solo (..)
+  (
+#if MIN_VERSION_ghc_prim(0,10,0)
+    Solo(MkSolo, Solo)
+#elif __GLASGOW_HASKELL__ >= 708 && \
+      __GLASGOW_HASKELL__ < 800 && \
+      defined(MIN_VERSION_OneTuple)
+# if MIN_VERSION_OneTuple(0,4,0)
+    Solo
+  , pattern Solo
+# else
+    Solo(Solo)
+# endif
+#else
+    Solo(Solo)
```

Comment:
> This is technically a breaking change, isn't it?

I don't think so. Here is my understanding of the situation:

* Prior to this PR, the `base-compat-batteries-0.12.*` series only supported building with `OneTuple-0.3.*`, which only exported a `Solo` data constructor. Moreover, GHC 9.4 itself only provided a `Solo` data constructor. As such, I don't think it was ever possible to use `MkSolo` via `base-compat-batteries-0.12.*` when building with GHC 9.4 or earlier.

  The one weird part is that GHC 9.6 provides both `MkSolo` and `Solo`, but I'm not proposing to change that, since I think it makes sense to match `base`'s API going forward (even at the expense of some weirdness in the CPP we use).
* After this PR, `base-compat-batteries` will still only export a `Solo` data constructor if building with GHC 9.4 or earlier, and it will export both `MkSolo` and `Solo` with GHC 9.6 or later.

As such, there shouldn't be any change in behavior. (Please correct me if I've overlooked something.)

> A comment mentioning that ghc-prim-0.10 is bundled with GHC-x.y (I don't remember myself when it happens) would make the conditionals easier to think about

Fair enough. I'll add some comments next to the CPP to clarify this.

Reply:
Oh I see, my bad!

I thought that `base-compat-batteries` were using `OneTuple-0.4` already!

Reply:
Sorry, this is rather confusing. The current major version of `base-compat-batteries` (`0.13.*`) _does_ support `OneTuple-0.4.*`, but it makes a deliberate effort to re-export both the `MkSolo` and `Solo` data constructors when possible. This PR, on the other hand, is making a change to the older `0.12.*` series of `base-compat-batteries`, which is done purely for the benefit of older build plans, such as the ones seen in #91.

Reply:
I see, thanks for explaining, I was indeed very confused.